### PR TITLE
feat: support resume_gateway_url for resuming

### LINKED ISSUE #762: Implement the new gateway resuming flow
discord/discord-api-docs#5282 shows the new way to resume a websocket, it is unclear how it actually works right now but it should be clearer soon.

### PR DESCRIPTION
**Pull request by GestaIt**
## Summary

Resolves #762 

Implements new gateway resume flow by using the `resume_gateway_url` from the `READY` event

## Checklist

- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: nextcord/gateway.py [hunk 1]
```diff
@@ -534,13 +540,15 @@ async def received_message(self, msg, /):
             self._trace = trace = data.get("_trace", [])
             self.sequence = msg["s"]
             self.session_id = data["session_id"]
+            self.resume_url = data["resume_gateway_url"]
             # pass back shard ID to ready handler
             data["__shard_id__"] = self.shard_id
             _log.info(
-                "Shard ID %s has connected to Gateway: %s (Session ID: %s).",
+                "Shard ID %s has connected to Gateway: %s (Session ID: %s). Reconnect specified as %s",
```
Comment (EmreTech) [inline nextcord/gateway.py:547]:
```suggestion
                "Shard ID %s has connected to Gateway: %s (Session ID: %s). Resume URL specified as %s",
```

**Review by EmreTech — APPROVED** [review]
LGTM

**Comment by ooliver1** [comment]
Did you make sure this complies with discord/discord-api-docs#5323

**Comment by GestaIt** [comment]
> Did you make sure this complies with [discord/discord-api-docs#5323](https://github.com/discord/discord-api-docs/pull/5323)

Yeah so `raise ReconnectWebSocket(self.shard_id, resume=False)` will mean that the resume url isn't used (line 534 of the revised gateway in this PR). However, it looks like there's an example where it resumed even after receiving op code 9 (line 528)? In that case, since resume is not specified to be false, it will attempt to use the resume url

In any event, the comment I left on line 703 of the revised client.py shows that when the resume keyword argument is false, we do not use the given resume url

I hope this clarified things!

**Comment by ooliver1** [comment]
Awesome, thank you!

**Review by ArjunSharda — APPROVED** [review]
LGTM

**Comment by ooliver1** [comment]
Unfortunately this is stalled due to this disappointing comment https://github.com/discord/discord-api-docs/pull/5282#issuecomment-1217108399

**Review by teaishealthy — APPROVED** [review]

**Comment by ooliver1** [comment]
Discord seem to have finalised, in a convoluted way, how invalid sessions are handled - the behaviour in discord/discord-api-docs#5182

**Review by ooliver1 — APPROVED** [review]

**Review by DenverCoder1 — APPROVED** [review]

**Pull request merged**